### DISCUSSION TITLE
Epiphan Type hardware compatibility

### DISCUSSION THREAD

**Discussion by Karl-Philippe** [PlusToolkit/PlusLib#920] [Q&A] [answered]
Hi,

I am trying to write the configuration file to aquire the frames of an ultrasound imaging system with a video grabber (Av.io from epiphane). I saw that the Epiphan frame grabber is supported by Plus (Type="Epiphan") but I was wondering if the Epiphan AV.io HD is supported by Plus (instead of the recommanded model for VGA and DVI capture: DVI2USB 3.0). I am using Plus Applications 2.8.0.2019-11-05 (Win64) on Windows.

Here is the Epiphan frame grabber configuration file template I am using:
```
<PlusConfiguration version="2.0">

<DataCollection StartupDelaySec="1.0" >
<DeviceSet
Name="PlusServer: Epiphan video capture device - color"
Description="Broadcasting acquired US video through OpenIGTLink"
/>
<Device
Id="VideoDevice"
Type="Epiphan" >
<DataSources>
<DataSource Type="Video" Id="Video" PortUsImageOrientation="MF" ImageType="RGB_COLOR" />
</DataSources>
<OutputChannels>
<OutputChannel Id="VideoStream" VideoDataSourceId="Video" />
</OutputChannels>
</Device>
<Device
Id="CaptureDevice"
Type="VirtualCapture"
BaseFilename="RecordingTest.igs.mha"
EnableCapturingOnStart="FALSE" >
<InputChannels>
<InputChannel Id="VideoStream" />
</InputChannels>
</Device>
</DataCollection>
<CoordinateDefinitions>
<Transform From="Image" To="Reference"
Matrix="
0.2 0.0 0.0 0.0
0.0 0.2 0.0 0.0
0.0 0.0 0.2 0.0
0 0 0 1" />
</CoordinateDefinitions>

<PlusOpenIGTLinkServer
MaxNumberOfIgtlMessagesToSend="1"
MaxTimeSpentWithProcessingMs="50"
ListeningPort="18944"
SendValidTransformsOnly="true"
OutputChannelId="VideoStream" >
<DefaultClientInfo>
<MessageTypes>
<Message Type="IMAGE" />
</MessageTypes>
<ImageNames>
<Image Name="Image" EmbeddedTransformToFrame="Reference" />
</ImageNames>
</DefaultClientInfo>
</PlusOpenIGTLinkServer>
</PlusConfiguration>
```

And I am getting the following error message:
```
|INFO|000.053000| Software version: Plus-2.8.0.62873a16 - Win64| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(122)
|INFO|000.053000| Logging at level 3 to file: C:/Users/kpbea/PlusApp-2.8.0.20191105-Win64/data/031522_172621_PlusLog.txt| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(123)
|INFO|000.570000| Supported devices: - 3dConnexion (ver: Plus-2.8.0) - AuroraTracker (ver: NDICAPI-1.7) - BkProFocusOem (ver: Plus-2.8.0) - ChRobotics (ver: Plus-2.8.0) - Epiphan (ver: Plus-2.8.0) - FakeTracker (ver: Plus-2.8.0) - GenericSerialDevice (ver: Plus-2.8.0) - ICCapturing (ver: The Imaging Source UDSHL-3.4) - ImageProcessor (ver: Plus-2.8.0) - IntelRealSense (ver: Plus-2.8.0) - Microchip (ver: Plus-2.8.0) - MmfVideo (ver: Plus-2.8.0) - NDITracker (ver: NDICAPI-1.7) - NoiseVideo (ver: Plus-2.8.0) - OpenCVVideo (ver: Plus-2.8.0) - OpenIGTLinkTracker (ver: OpenIGTLink v3.1.0) - OpenIGTLinkVideo (ver: OpenIGTLink v3.1.0) - OptiTrack (ver: Plus-2.8.0) - OpticalMarkerTracker (ver: Plus-2.8.0) - PhidgetSpatial (ver: Plus-2.8.0) - PolarisTracker (ver: NDICAPI-1.7) - SavedDataSource (ver: Plus-2.8.0) - UsSimulator (ver: Plus-2.8.0) - VFWVideo (ver: Plus-2.8.0) - VirtualBufferedCapture (ver: Plus-2.8.0) - VirtualCapture (ver: Plus-2.8.0) - VirtualDiscCapture (ver: Plus-2.8.0) - VirtualMixer (ver: Plus-2.8.0) - VirtualSwitcher (ver: Plus-2.8.0) - VirtualTextRecognizer (ver: Plus-2.8.0) - VirtualVolumeReconstructor (ver: Plus-2.8.0) - WitMotionTracker (ver: Plus-2.8.0) | in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(127)
|INFO|000.578000| Server host name: PC-KarlPhilippe| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(158)
|INFO|000.589000| Server IP addresses: [169.254.167.170](https://169.254.167.170/), [169.254.96.22](https://169.254.96.22/), [169.254.249.192](https://169.254.249.192/), [169.254.199.87](https://169.254.199.87/), [169.254.204.47](https://169.254.204.47/), [172.29.2.46](https://172.29.2.46/), [169.254.73.82](https://169.254.73.82/), 127.0.0.1| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(178)
|INFO|000.589000| Start remote control server at port: 18904| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(187)
|INFO|005.327000| Connect using configuration file: C:\Users\kpbea\Downloads\SlicerIGT-Data\PlusDeviceSet_Server_EpiphanVideoCapture.xml| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(639)
|INFO|005.328000| Server process command line: "C:/Users/kpbea/PlusApp-2.8.0.20191105-Win64/bin/PlusServer.exe" --config-file="PlusDeviceSet_Server_EpiphanVideoCapture.xml" --verbose=3| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(320)
|ERROR|005.580000| Server stopped unexpectedly. Return code: 1| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(864)
|INFO|005.594000| Disconnect request successful| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(632)
|ERROR|005.595000| Failed to start server process| in E:\D\PSNP64b\PlusApp\PlusServerLauncher\PlusServerLauncherMainWindow.cxx(338)
```

Thank you,

**Comment by lassoan** [comment, marked as answer]
AV.io HD does not use the Epiphan drivers or SDK, so you cannot use the Epiphan device to connect to it. Instead you should be able to access it using a generic imaging device, such as Microsoft Media Foundation or OpenCV.

**Reply by Karl-Philippe** [reply]
Thank you for your answer @lassoan! It worked fine in the Win64 version but I have to use the Win32 version since I use the Ascension NDI tracking system in parallel (in order to aquire tracked images). Even if it is indicated in the documentation that Microsoft Media Foundation is compatible with the Win32 version I had errors that I did not have in the Win64 version. I am using the MMF configuration file template (which works fine with the Win64 version) and here is the error I am getting in the Win32 version:
`|ERROR|412.555000|SERVER> Source Reader error: 80070005| in :\D\PSNPb\PlusLib\src\PlusDataCollection\MicrosoftMediaFoundation\vtkPlusMmfVideoSource.cxx(145)`

Could you please explain me the meaning of this error?

**Reply by lassoan** [reply]
Err where the error is logged:

https://github.com/PlusToolkit/PlusLib/blob/7d0081f7190b6f341c729f2ef5b48b5f95b2254a/src/PlusDataCollection/MicrosoftMediaFoundation/vtkPlusMmfVideoSource.cxx#L145

It means that the image acquisition pipeline is set up but when a new frame is supposed to be acquired there is an error. It suggests that the driver only supports 64-bit applications.

You can use a 64-bit PlusServer for image acquisition and start a 32-bit PlusServer for interfacing with the tracker. You can stream the acquired transforms from the 32-bit every to the 64-bit server via OpenIGTLink.

**Reply by Karl-Philippe** [reply]
Ok thank you very much I did knew it was possible. So if I understand correctly, there is no need to use Virtual Mixer to combine transformations and images in the configuration files anymore and it is mixed via OpenIGTLink? To give some context my application is a tracked ultrasoud system

**Reply by lassoan** [reply]
I would recommend to use the virtual mixer device in the 64-bit server to make the time stamping more consistent (transforms are updated exactly at the same time the image is updated).

**Reply by Karl-Philippe** [reply]
Perfect, thank you! However, I didn't see in the documentation how to import a TrackingStream (or a Stream in general) from another Plus server, do I have to create a virtual device and then use the VirtualMixer or is it possible to import the TrackingStream directly from the first server? I'm not sure to understand what the output of the Plus server of the tracking stream and the input of the image acquisition and data mixing Plus server are.
Also, do I need to change the ListeningPort of one of the Plus server, so they are different? CurrentlyI have the typical value of ListeningPort="18944" for both.

**Reply by lassoan** [reply]
Here is an example configuration file to receive transforms via OpenIGTLink from another PlusServer: http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkTracker.html

**Reply by Karl-Philippe** [reply]
Thanks @lassoan this is what I was looking for! Both servers (trackingStream and TrackedVideoStream) are now up and running without errors however as soon as I try to aquire the trackedVideoStream server from OpenIGTLinkIF in Slicer (set the IGTConnector status to "active") I get periodicaly the for following error messages:
```
|ERROR|206.140000|SERVER> Failed to get video buffer item UID from time: 194.542243| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\vtkPlusChannel.cxx(1425)
|ERROR|206.141000|SERVER> Unable to get most recent timestamp!| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\vtkPlusChannel.cxx(754)
|ERROR|206.141000|SERVER> Failed to get tracked frame list from data collector (last recorded timestamp: 189.641857| in :\D\PSNP64b\PlusLib\src\PlusServer\vtkPlusOpenIGTLinkServer.cxx(432)
|INFO|206.225000|SERVER> OpenIGTLink broadcasting started. No data was available between 189.642-189.675sec, therefore no data were broadcasted during this time period.
|INFO|206.326000|SERVER> OpenIGTLink broadcasting started. No data was available between 189.775-189.776sec, therefore no data were broadcasted during this time period.
|INFO|206.459000|SERVER> OpenIGTLink broadcasting started. No data was available between 189.876-189.909sec, therefore no data were broadcasted during this time period.
|INFO|206.560000|SERVER> OpenIGTLink broadcasting started. No data was available between 190.009-190.009sec, therefore no data were broadcasted during this time period.
```

I also get this error message intermittently when I start the TrackedVideoStream Plus server:
```
|ERROR|012.689000|SERVER> Failed to get video buffer item UID from time: 1.113957| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\vtkPlusChannel.cxx(1425)
```

Could you please explain me the meaning of those errors?

**Reply by lassoan** [reply]
> |ERROR|206.140000|SERVER> Failed to get video buffer item UID from time: 194.542243| in :\D\PSNP64b\PlusLib\src\PlusDataCollection\vtkPlusChannel.cxx(1425)

This error is logged because current time (time elapsed since application startup) is 206.14sec and the latest tracker timestamp is 194.54sec - more than 10sec old. There is no such old video frames available in the buffer, which is understandable.

The question is why the latest tracker data is so old. If it is due to the PlusServer not sending any data if there is no valid data (e.g., tool out of view) then all you need is to enable sending invalid transforms (`SendValidTransformsOnly="false"`).

If the problem persists then probably the timestamps are not synchronized. If you acquire data on two different computers then you either need to synchronize their clocks (for more robust timestamping) or [enable `UseReceivedTimestamps`](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkTracker.html).

**Reply by Karl-Philippe** [reply]
Thank you for your answer @lassoan! I am using a single windows computer for both servers. Finally, I solved the previous errors by setting the Device's (acquiring the tracking server) MessageType to "TRANSFORM" (instead of "TDATA")  but I am getting the following warning message:
```
|WARNING|6591.361000|SERVER> Incoming TRANSFORM message was received without a tool status. Defaulting to TOOL_OK.| in :\D\PSNP64b\PlusLib\src\PlusOpenIGTLink\vtkPlusIgtlMessageCommon.cxx(795)
```

Is this something to worry about?

Thank you,

Here are my configuration files:
[PlusDeviceSet_Server_Ascension3DG.txt](https://github.com/PlusToolkit/PlusLib/files/8341153/PlusDeviceSet_Server_Ascension3DG.txt)
[PlusDeviceSet_Server_MmfColorVideoCapture+Tracking.txt](https://github.com/PlusToolkit/PlusLib/files/8341154/PlusDeviceSet_Server_MmfColorVideoCapture%2BTracking.txt)

**Reply by lassoan** [reply]
Since Ascension tracker sensors practically never reported to get out of view, you don't need to worry about this warning.

**Reply by Karl-Philippe** [reply]
Ok perfect! Thanks a lot @lassoan!
